Updated Roboto font to download from google on the web rather than being...

PaulL1 opened the pull request:
Issue #121 resolution - as described by Josh, add direct inclusion of Roboto from fonts.googleapis.com, and remove the fontface definition from main.less.

I have completed basic testing - the base ng-boilerplate now correctly uses Roboto without any further tailoring.

PaulL1 committed:
Updated Roboto font to download from google on the web rather than being local

joshdmiller commented:
Nice. Thanks. :-)

PaulL1 committed:
Added protractor and a sample scenario, along with ids in the index page to make testing easier, and changes to .gitignore and the documentation

PaulL1 committed:
Fixes to the tests

PaulL1 commented:
Unfortunately github issues are getting the better of me!!  Anyway, I didn't really mean to add this commit on top of the previous commit, but I appear to have done so.

This commit adds in protractor, as described in #196, and I'll add some further commentary there about what's missing.  I think we're going to have to refactor/reorganise to apply to 0.4 anyway, so I'm not going to fight github to make the commit separate at this stage.

joshdmiller commented:
@PaulL1 In the future, feature branches are the way to solve this very common issue. Each new feature you work on stays in its own git branch; the PR is then just a request to merge the change (or changes) from a specific feature branch into the main branch (or even to an upstream branch).

But your instinct about 0.4 is right, so don't worry about it. If you're curious, the new NGBP (less many features that I haven't pushed yet) is on the [`0.4wip` branch](https://github.com/ngbp/ng-boilerplate/tree/v0.4-wip), which is not to be confused with the old one, which has a hyphen in its name. 

ngbp 0.4 is just a runner with some management features; it registers a single multitask called `ngbp` which has subtasks responsible for doing stuff, like `ngbp:build` and `ngbp:compile` as well as managing dependencies more intelligently.

Speaking of dependencies, all functionality has been moved to plugin packages that are themselves separate repos. I have only pushed [`ngbp-contrib-clean`](https://github.com/ngbp/ngbp-contrib-clean) to GitHub so far. These are each super simple: they load in their dependent grunt tasks from NPM (in this case `grunt-contrib-clean`), define a default configuration, and then tell ngbp when to run which tasks. This is the entire thing:

``` js
var ngbp = require( 'ngbp' );

module.exports = function ( grunt ) {
  grunt.loadNpmTasks( 'grunt-contrib-clean' );

  var task = ngbp.task.create( 'clean', {
    ngbp_build: [
      '<%= ngbp.build_dir %>'
    ],
    ngbp_compile: [
      '<%= ngbp.compile_dir %>'
    ]
  });

  task.run( 'ngbp_build', 'prebuild', 80 );
  task.run( 'ngbp_compile', 'precompile', 80 );
};
```

Those `task.run` calls put the specified task (e.g. `copy:ngbp_build`) into a certain hook (e.g. `prebuild`) with a specified priority (e.g. `80`). All the `ngbp:build` task does is run first a set of prebuild tasks, then a set of build tasks, and finally a set of postbuild tasks. To run each of those hooks, it sorts the tasks specified by the priority, which is user-customizable.

This makes ngbp much more customizable as a whole. Installing ngbp is just to run `$ npm install ngbp` and creating a barebones `Gruntfile.js` that runs `grunt.loadNpmTasks('ngbp');`. Then you can install these third-party modules by simply enabling them, e.g.:

``` sh
$ grunt ngbp:use:ngbp-contrib-less
```

This handles dependencies for you, installing the requisite npm module and ensuring it will get loaded during builds, thus injecting the appropriate tasks into the right hooks with their default priorities.

There's a lot more to 0.4, but since you didn't even ask for the veritable deluge of info I just threw at you, I'll stop here. :-)

So yeah, don't worry about the details of the PRs. They'll be great for discussion, but well have to port them to 0.4 as you thought.

PaulL1 committed:
Exclude the scenario files (protractor) from grunt

PaulL1 committed:
Fixed typos in grunt file, added grunt-protractor-runner and a new grunt task of grunt e2e

PaulL1 committed:
Added astrolabe to allow modularity in tests

PaulL1 committed:
Updated doco too

PaulL1 commented:
Refer again to the commentary on #196 as to how to use astrolabe etc.

PaulL1 committed:
Added the e2e test files to grunt watch

PaulL1 committed:
Modified gruntfile to change 'karma:continuous' to 'karma:runonce', as that is what it actually seems to do, and to split build and build_only, so that karma:runonce is executed on a build, but not on a watch, which for some reason I don't understand breaks watch

PaulL1 commented:
OK, that last change is nothing to do with e2e testing, just a general fix to the gruntfile, reflecting issue #37.  I was trying to work out what was going on, and my conclusion was:
1.  The task labelled karma:continuous was actually karma:runonce - because it explicitly ran once.  I've changed the name to make that clear
2.  People were removing the old karma:continuous (now karma:runonce) from the build task.  That fixed their problem with watch, but broke compile and build from doing anything sensible

To that end, I've refactored.  I've created a "build_only" task, which is included in watch. This task doesn't include the old karma:continuous (now karma:runonce).  I then changed build to call build_only, then karma:runonce, so the old build and compile behaviours remain without breaking watch.

This clearly would be much more useful as a feature commit.  Unfortunately having messed up the 3.1 branch in my fork, I have no way to easily create that.  So perhaps this is a 0.4 feature as well.  Those who want this standalone could pick out these changes from the commit diff.
